log/scope.go

```diff
@@ -315,3 +338,96 @@ func (s *Scope) SetLogCallers(logCallers bool) {
 func (s *Scope) GetLogCallers() bool {
 	return s.logCallers.Load().(bool)
 }
+
+// Copy makes a copy of s and returns a pointer to it.
+func (s *Scope) Copy() *Scope {
+	out := *s
+	out.context = copyStringInterfaceMap(s.context)
+	out.labels = copyStringInterfaceMap(s.labels)
+	return &out
+}
+
+// WithLabels adds a key-value pairs to the labels in s.
+func (s *Scope) WithLabels(kvlist ...interface{}) *Scope {
+	out := s.Copy()
+	if len(kvlist)%2 != 0 {
+		out.labels["WithLabels error"] = fmt.Sprintf("even number of parameters required, got %d", len(kvlist))
+		return out
+	}
+
+	for i := 0; i < len(kvlist); i += 2 {
+		keyi := kvlist[i]
+		key, ok := keyi.(string)
+		if !ok {
+			out.labels["WithLabels error"] = fmt.Sprintf("label name %v must be a string, got %T ", keyi, keyi)
+			return out
+		}
+		out.labels[key] = kvlist[i+1]
+		out.labelKeys = append(out.labelKeys, key)
+	}
+	return out
+}
+
+// WithoutLabels makes a copy of s, clears labels in s with the given keys and returns the copy.
+// Not-existent keys are ignored.
+func (s *Scope) WithoutLabels(keys ...string) *Scope {
+	out := s.Copy()
+	for _, key := range keys {
+		delete(out.labels, key)
+		out.labelKeys = removeKey(out.labelKeys, key)
+	}
+	return out
+}
+
+// WithoutAnyLabels clears all labels from a copy of s and returns it.
+func (s *Scope) WithoutAnyLabels() *Scope {
+	out := s.Copy()
+	out.labels = make(map[string]interface{})
+	out.labelKeys = nil
+	return out
+}
+
+// callHandlers calls all handlers registered to s.
+func (s *Scope) callHandlers(
+	severity Level,
+	scope *Scope,
+	ie *structured.Error,
+	msg string,
+	fields []zapcore.Field) {
+	for _, h := range defaultHandlers {
+		h(severity, scope, ie, msg, fields)
+	}
+}
+
+// getErrorStruct returns (*Error, 1) if it is the first argument in the list is an Error ptr,
+// or (nil,0) otherwise. The second return value is the offset to the first non-Error field.
+func getErrorStruct(fields ...interface{}) (*structured.Error, int) {
```

Comment:
nit: why even return int here? seems like callers can just check if the response is == nil. Its internal though so NBD

Reply:
just to avoid checking every time at the caller site. makes the code more compact for the 50 wrapper functions.

Reply:
ack, thanks